### PR TITLE
Added a new closable driver pool.

### PR DESCRIPTION
**Pull request by Matt-Guest**
To avoid breaking builds a new closable interface was created
with a new create function.

I also added some thread guards around the close function and open
function.

### CODE REVIEW

**Comment by Matt-Guest** [comment]
I been using the following code to test goroutines against the changes
```package main

import (
	"fmt"
	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
	"time"
)

func main() {
	pool, _ := bolt.NewClosableDriverPool("bolt://localhost:7687", 5)
	go open(pool)
	go open(pool)
	go open(pool)
	go open(pool)
	go open(pool)
	go open(pool)
	time.Sleep(time.Millisecond * 20)
	err := pool.Close()
	go open(pool)
	go open(pool)
	time.Sleep(time.Millisecond * 200)
	fmt.Print("Closed\n")
	if err != nil {
		fmt.Printf("Pool err %s\n", err.Error())
	}

}

func open(pool bolt.ClosableDriverPool) {
	conn, err := pool.OpenPool()
	if err != nil {
		fmt.Print("conn err %s \n", err.Error())
		return
	}
	fmt.Print("conn Ok \n")
	result, err := conn.ExecNeo("CREATE (n:NODE {foo: {foo}, bar: {bar}})", map[string]interface{}{"foo": 1, "bar": 2.2})
	if err != nil {
		fmt.Printf("Conn err %s\n", err.Error())
		return
	}
	numResult, _ := result.RowsAffected()
	fmt.Printf("CREATED ROWS: %d\n", numResult) // CREATED ROWS: 1
	conn.Close()

}
```

I tried running there tests but they fail. I ran the same tests on `master` and i get the same test failures. So at the moment the tests are not giving me any feedback if anything else is broken.

File: driver.go [hunk 1]
```diff
@@ -73,14 +75,33 @@ type DriverPool interface {
 	reclaim(*boltConn) error
 }
 
+// ClosableDriverPool like the DriverPool but a closable function
```
Comment (daiLlew) [inline driver.go]: "CloseableDriverPool like the DriverPool but not half baked"

File: driver.go [hunk 2]
```diff
@@ -101,13 +122,38 @@ func NewDriverPool(connStr string, max int) (DriverPool, error) {
 
 // OpenNeo opens a new Bolt connection to the Neo4J database.
 func (d *boltDriverPool) OpenPool() (Conn, error) {
-	conn := <-d.pool
-	if conn.conn == nil {
-		if err := conn.initialize(); err != nil {
-			return nil, err
+	// For each connection request we need to block in case a the Close function is called. This gives us a guarantee
+	// when closing the pool no new connections are make.
+	d.refLock.Lock()
```
Comment (daiLlew) [inline driver.go]: Will this block here if the lock is already obtained?

File: driver.go [hunk 3]
```diff
@@ -101,13 +122,38 @@ func NewDriverPool(connStr string, max int) (DriverPool, error) {
...
+	defer d.refLock.Unlock()
+	if !d.closed {
+		conn := <-d.pool
+		if conn.conn == nil {
+			if err := conn.initialize(); err != nil {
+				return nil, err
```
Comment (daiLlew) [inline driver.go]: Do you need to add the conn back to the buffer if this fails?

File: driver.go [hunk 4]
```diff
@@ -3,6 +3,8 @@ package golangNeo4jBoltDriver
 import (
 	"database/sql"
 	"database/sql/driver"
+	"sync"
+	"github.com/johnnadratowski/golang-neo4j-bolt-driver/errors"
```
Comment (daiLlew) [inline driver.go]: do you need the full import path here? Can you not use a the relative path within the context of the project?

File: driver.go [hunk 5]
```diff
@@ -101,13 +122,38 @@ func NewDriverPool(connStr string, max int) (DriverPool, error) {
...
+// Close all connections in the pool
+func (d *boltDriverPool) Close() error {
+	var err error
+	// Lock the connection ref so no new connections can be added
+	d.refLock.Lock()
+	defer d.refLock.Unlock()
+	for _, conn := range d.connRefs {
+		// Remove the reference to the pool to allow a clean up of the connection
+		conn.poolDriver = nil
+		err = conn.Close()
```
Comment (daiLlew) [inline driver.go]: You are potentially overwriting each error here.

File: driver.go [hunk 6]
```diff
@@ -101,13 +122,38 @@ func NewDriverPool(connStr string, max int) (DriverPool, error) {
...
 	}
-	return conn, nil
+	// Mark the pool as closed to stop any new connections
+	d.closed = true
```
Comment (daiLlew) [inline driver.go]: Maybe defer this as well?

`defer func() { d.refLock.Unlock() d.closed }()`

File: driver.go (hunk 2 above)
Comment (Matt-Guest) [reply driver.go]: Yes, as we don't want to create and close at the same time. Doing so could lead to new connections not being closed.

File: driver.go (hunk 4 above)
Comment (Matt-Guest) [reply driver.go]: Its need as the standard go library has an error package. We should try keep to the authors style.

File: driver.go (hunk 3 above)
Comment (Matt-Guest) [reply driver.go]: The connection is returned when closed. The current version will still leak connections if they are not closed

File: driver.go (hunk 5 above)
Comment (Matt-Guest) [reply driver.go]: Yes I am 😞 , I fix this.

File: driver.go (hunk 6 above)
Comment (Matt-Guest) [reply driver.go]: I address this too

File: driver.go [hunk 7]
```diff
@@ -73,14 +75,33 @@ type DriverPool interface {
 	reclaim(*boltConn) error
 }
 
+// ClosableDriverPool like the DriverPool but with a closable function
+type ClosableDriverPool interface {
+	OpenPool() (Conn, error)
```
Comment (ian-kent) [inline driver.go]: could embed the DriverPool interface to avoid duplication?

File: driver.go (hunk 6 above)
Comment (ian-kent) [reply driver.go]: not clear from the suggestion, but to clarify, d.closed would need to be set to true before the mutex is unlocked

File: driver.go (hunk 5 above)
Comment (Matt-Guest) [reply driver.go]: I had a quick read about returning multiple error and came to the conclusion if any of the connections failed they will all fail. So I made it return on the first error found.

File: driver.go (hunk 6 above)
Comment (Matt-Guest) [reply driver.go]: I added the defer in but I kept setting the boolean at the bottom. My reasons for this is to keep the flow more natural.

**Pull request merged**